### DISCUSSION TITLE
Problem linking against inet while dockerizing

### DISCUSSION THREAD

**Discussion by phdavis1027** [inet-framework/inet#955] [General]
I am trying to build a Docker image from the tag `omnetpp/inet:o6.0-4.4.1`. Once I move model into the container, I run the executable generated by opp_makemake's makefile with the flag `-l $INET_ROOT/src/INET`, but I get an error message saying that no `libINET.so` can be found. Any insight would be appreciated, I'm still very confused by the omnetpp CLI.

Best,
Phillip

**Comment by rhornig** [comment]
If you have an executable that uses the libINET.so as a shared library you must either:
- Add the `<inet_root>/src` directory to the LD_LIBRARY_PATH environment variable (i.e. the location of the library).
- Or during the linking phase of your executable, you must provide a `-Wl,-rpath,<omnetpp_root>/src` option to add the location of the lib to the runtime path.

**Reply by phdavis1027** [reply]
When you say "the linking phase of our executable," is that something that I give to `opp_makemake`? I'm a bit confused about the behavior of the executables produced by omnet CLI, but that might be a result of the codebase I'm left with.

**Reply by rhornig** [reply]
"Once I move model into the container" indicates that you are building your executable on your system and then copy the finished executable into a docker image and try to run it from there?

The OMNeT++ binaries are NOT relocatable, meaning that all executables (shared object or exe) MUST be exactly at the same directory location where it was originally built. I suspect that the location of the inet shared lib is different on your system than on the container. You may try to experiment with the LD_LIBARAY_PATH approach, but the proper solition is to create a Docker file that actually build your executable INSIDE the container.

**Reply by phdavis1027** [reply]
Even when I build in the container I get an error saying that it can't resolve a particular kind of link extended from Etherlink. It's successfully loading NEDs from $INET_ROOT/src/inet and I'm running `./models -l $INET_ROOT/src/INET`, so it should be linking properly.

EDIT: I found the source of the issue. Now I'm fairly sure I'm just trying to port the code to the newer version of Omnet/Inet that ships with the container image.

**Reply by phdavis1027** [reply]
Incidentally, wouldn't it make more sense to install the inet .so in a system directory, or at least in a local directory called `lib`? I would not normally look for library files in a directory called `src`

**Reply by rhornig** [reply]
Yeah, for a normal program, installing it in the system lib folder would be the right choice but OMNeT++ and INET is not a fixed tool, but something the end user constantly tweak, recompile modify (at least in most cases) so we keep everything in the source tree and avoid system wide changes (installing into lib). Because of this we have to manually fiddle with the library path (both link and runtime). This is bad, but we are mostly hiding it from the user and much better than requiring the user to do `sudo make install` on each model change.

I have to admit the location of `src` for the final executable is ugly but it has some historical reasons (in INET). Someday we might change that.

**Reply by rhornig** [reply]
Oh, one shortcut to get it working: have you tried using [opp_env](https://github.com/omnetpp/opp_env)? That tool allows you to setup any working combination of omnet/inet. There is also a docker image that contains a pre-installed opp_env tool so you can build on top of that to create your own

**Reply by phdavis1027** [reply]
Thanks for the suggestion, we might look into that for future use (for example if our project lives long enough that we wind up having to support multiple versions) but I wound up just reading through the code and giving specific indices into the gate arrays.
